File: 0015-3sum/0015-3sum.rs
impl Solution {
    pub fn three_sum(numbers: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = numbers.clone();
        let mut res = Vec::new();
        
        /// sort the input array
        nums.sort();

        
        for i in 0..nums.len() {
            // From now on only > 0 can be found, so no need to iterate further
            if nums[i] > 0 {
                return res;
            }
            
            //skip all identical entries to avoid duplicate answers
            if i > 0 && nums[i] == nums[i - 1] {
                continue;
            }

            
            let mut left = i + 1;
            let mut right = nums.len() - 1;

            while left < right {
                let sum = nums[i] + nums[left] + nums[right];
                if sum > 0 {
                    right -= 1;
                } else if sum < 0 {
                    left += 1
                } else {
                    res.push(vec![nums[i], nums[left], nums[right]]);
                    left += 1;

                    while nums[left] == nums[left - 1] && left < right {
                        left += 1;
                    }
                }
            }
        }

        return res;
    }
}

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        
        std::sort(nums.begin(), nums.end()); ///O(Nlog(N))
        std::vector<std::vector<int>> res;
        for (int i = 0; i < n; ++i) {
            if (nums[i] > 0) {
                ///the chance of finding a triplet of elements that summed up to 0 is impossible at this point
                break;
            }
            
            int l = i + 1,
                r = n - 1;
            
            while (l < r) {
                if (nums[i] + nums[l] + nums[r] == 0) {
                    std::vector<int> temp = {nums[i],nums[l],nums[r]};
                    res.push_back(temp);
                    
                    /// skip all r duplicates
                    while (l <= r && nums[r] == temp[2]) {
                        --r;
                    }
                    
                    /// skip all l duplicates
                    while (l <= r && nums[l] == temp[1]) {
                        ++l;
                    }
                } else if (nums[i] + nums[l] + nums[r] > 0) {
                    --r;
                } else {
                    ++l;
                }
            }
            while(i+1 < nums.size() && nums[i]==nums[i+1]) i++;
        }
        
        return res;
    }
};